### PR TITLE
Ticket 44 : Search table

### PR DESCRIPTION
**Pull request by maximFeldman1**
create search table and used library lodash debounce for good performance

### CODE REVIEW

File: src/Components/Actions/SearchTable/index.tsx [hunk 1]
```diff
@@ -1,10 +1,24 @@
-import React from "react";
+import React, { useCallback, useEffect, useState } from "react";
 import { InputText } from "primereact/inputtext";
+import { useProductContext } from "../../../context/contextProducts";
+import debounce from "lodash.debounce";
 
 export const SearchTable = () => {
+  const { text, setText } = useProductContext();
+
+  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (itairom-appsforce) [inline src/Components/Actions/SearchTable/index.tsx]: wrap it with useCallback

File: src/Components/Actions/SearchTable/index.tsx [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+    setText(e.target.value);
+  };
+
+  const debouncedChangeHandler = debounce(changeHandler, 1000);
```
Comment (itairom-appsforce) [inline src/Components/Actions/SearchTable/index.tsx]: maybe
```suggestion
  const onChangeHandler = debounce(changeHandler, 1000);
```

File: src/Components/Actions/SearchTable/index.tsx [hunk 3]
```diff
@@ -1,10 +1,24 @@
...
+
   return (
     <div>
-      <InputText type="text" className="block " placeholder="type..." />
+      <InputText
+        type="text"
+        onChange={debouncedChangeHandler}
+        placeholder="type..."
```
Comment (itairom-appsforce) [inline src/Components/Actions/SearchTable/index.tsx]: add placeholder to i18

File: src/context/contextProducts/index.tsx [hunk 4]
```diff
@@ -1,23 +1,27 @@
...
 }
 
 export const ProductContext = createContext<IProps>({ products });
 
 export const ProductContextProvider = ({ children }: any) => {
-  const { data: products, refetch } = useQuery(
-    "getAllProducts",
-    ProductService.getAll
-  );
+  const [text, setText] = useState<string | "">("");
```
Comment (itairom-appsforce) [inline src/context/contextProducts/index.tsx]: thats work?
```suggestion
  const [text, setText] = useState<string>("");
```

Comment (itairom-appsforce) [inline src/context/contextProducts/index.tsx]: also 
```suggestion
  const [search, setSearch] = useState<string | "">("");
```

File: src/mockServer/controllers/products.ts [hunk 5]
```diff
@@ -6,7 +6,17 @@ import { IProduct } from "models";
 export class ProductController {
   static getAll(req: RestRequest, res: ResponseFunction, ctx: RestContext) {
     try {
-      return res(ctx.status(200), ctx.json(products));
+      const searchText: string | null = req.url.searchParams.get("search");
+      let filterProducts;
+      if (searchText !== null) {
```
Comment (itairom-appsforce) [inline src/mockServer/controllers/products.ts]:
```suggestion
      if (!searchText) {
```

File: src/Components/Actions/SearchTable/index.tsx (hunk 1 above)
Comment (maximFeldman1) [reply src/Components/Actions/SearchTable/index.tsx]: Fixed

File: src/Components/Actions/SearchTable/index.tsx (hunk 2 above)
Comment (maximFeldman1) [reply src/Components/Actions/SearchTable/index.tsx]: Fixed

File: src/context/contextProducts/index.tsx (hunk 4 above)
Comment (maximFeldman1) [reply src/context/contextProducts/index.tsx]: Fixed

File: src/mockServer/controllers/products.ts (hunk 5 above)
Comment (maximFeldman1) [reply src/mockServer/controllers/products.ts]: its should be  -> if (searchText){}

File: src/Components/Actions/SearchTable/index.tsx (hunk 3 above)
Comment (maximFeldman1) [reply src/Components/Actions/SearchTable/index.tsx]: fixed

File: src/Components/Actions/SearchTable/index.tsx [hunk 6]
```diff
@@ -1,10 +1,29 @@
-import React from "react";
+import React, { useCallback, useEffect, useState } from "react";
 import { InputText } from "primereact/inputtext";
+import { useProductContext } from "../../../context/contextProducts";
+import debounce from "lodash.debounce";
+import { useTranslation } from "react-i18next";
 
 export const SearchTable = () => {
+  const { text, setText } = useProductContext();
```
Comment (amitmarko) [inline src/Components/Actions/SearchTable/index.tsx]:
```suggestion
  const { search, onSearchChange } = useProductContext();
```

File: src/context/contextProducts/index.tsx [hunk 7]
```diff
@@ -1,23 +1,28 @@
...
 }
 
 export const ProductContext = createContext<IProps>({ products });
 
 export const ProductContextProvider = ({ children }: any) => {
-  const { data: products, refetch } = useQuery(
-    "getAllProducts",
-    ProductService.getAll
-  );
+  const [text, setText] = useState<string>("");
```
Comment (amitmarko) [inline src/context/contextProducts/index.tsx]:
```suggestion
  const [search, setSearch] = useState<string>("");
```

File: src/context/contextProducts/index.tsx [hunk 8]
```diff
@@ -1,23 +1,28 @@
...
+
+  const { data: products, refetch } = useQuery({
+    refetchOnWindowFocus: false,
+    queryKey: ["text", text],
+    queryFn: () => ProductService.getAll({ search: text }),
+  });
 
-  const value = { products, refetch };
+  const value = { products, refetch, setText, text };
```
Comment (amitmarko) [inline src/context/contextProducts/index.tsx]:
```suggestion
  const value = { products, refetch, onSearchChange: handleSerachChange, serach };
```

File: src/i18n/translation/en.ts [hunk 9]
```diff
@@ -2,9 +2,11 @@ import { common } from "./common/en";
 import { toast } from "./products/toast/en";
```
Comment (amitmarko) [inline src/i18n/translation/en.ts:2]: translation files should be separate by page names + common

**Pull request merged**